hal/src/rtl872x/gpio_hal.cpp

```diff
@@ -433,3 +434,64 @@ uint32_t hal_gpio_pulse_in(hal_pin_t pin, uint16_t value) {
 
     return (hal_timer_micros(nullptr) - pulse_start);
 }
+
+int hal_gpio_get_drive_strength(hal_pin_t pin, hal_gpio_drive_t* drive) {
+    uint32_t rtlPin = hal_pin_to_rtl_pin(pin);
+
+    /* get PADCTR */
+    uint32_t drvStrength = PINMUX->PADCTR[rtlPin];
+
+    /* get Pin_Num drvStrength contrl */
+    drvStrength &= PAD_BIT_MASK_DRIVING_STRENGTH << PAD_BIT_SHIFT_DRIVING_STRENGTH;
+    drvStrength >>= PAD_BIT_SHIFT_DRIVING_STRENGTH;
+
+    // Pad driving strength
+    //   - PAD_DRV_STRENGTH_0: 4mA
+    //   - PAD_DRV_STRENGTH_1: 8mA   (Severe overshoot, not recommended to use it)
+    //   - PAD_DRV_STRENGTH_2: 12mA
+    //   - PAD_DRV_STRENGTH_3: 16mA  (Severe overshoot, not recommended to use it)
+    switch(drvStrength) {
+        case PAD_DRV_STRENGTH_0: *drive = HAL_GPIO_DRIVE_STANDARD; break;
+        case PAD_DRV_STRENGTH_2: *drive = HAL_GPIO_DRIVE_HIGH; break;
+        // DVOS won't configure PAD_DRV_STRENGTH_1 and PAD_DRV_STRENGTH_3 due to the
+        // severe overshoot, so it is the default setting
+        default:
+            *drive = HAL_GPIO_DRIVE_DEFAULT; break;
+    }
+
+    return SYSTEM_ERROR_NONE;
+}
+
+int hal_gpio_configure_drive_strength(hal_pin_t pin, hal_gpio_drive_t drive) {
+    uint32_t rtlPin = hal_pin_to_rtl_pin(pin);
+    uint32_t drvStrength = PAD_DRV_STRENGTH_0;
+
+    // Pad driving strength
+    //   - PAD_DRV_STRENGTH_0: 4mA
+    //   - PAD_DRV_STRENGTH_1: 8mA   (Severe overshoot, not recommended to use it)
+    //   - PAD_DRV_STRENGTH_2: 12mA
+    //   - PAD_DRV_STRENGTH_3: 16mA  (Severe overshoot, not recommended to use it)
+    switch(drive) {
+        case HAL_GPIO_DRIVE_STANDARD: drvStrength = PAD_DRV_STRENGTH_0; break;
+        case HAL_GPIO_DRIVE_HIGH:     drvStrength = PAD_DRV_STRENGTH_2; break;
+        case HAL_GPIO_DRIVE_DEFAULT:  drvStrength = PAD_DRV_STRENGTH_3; break;

```

Comment:
Using `HAL_GPIO_DRIVE_DEFAULT` will set `PAD_DRV_STRENGTH_3`? I thought we did not want to use this mode? 
In your opinion, what should the default drive strength be? `PAD_DRV_STRENGTH_0`?

Reply:
@scott-brust  Realtek defaults to using PAD_DRV_STRENGTH_3 as the default drive strength, as indicated in the log. The overshoot issue occurs on high-speed buses, and it has been resolved in each peripheral driver. However, considering the possibility of introducing other side effects (which might be an unnecessary concern), I have decided to set the default drive strength to `PAD_DRV_STRENGTH_3`.
```
hal_gpio_get_drive_strength, pin: D0, drvStrength: 3
hal_gpio_get_drive_strength, pin: D1, drvStrength: 3
hal_gpio_get_drive_strength, pin: D2, drvStrength: 3
...
hal_gpio_get_drive_strength, pin: D18, drvStrength: 3
hal_gpio_get_drive_strength, pin: D19, drvStrength: 3
hal_gpio_get_drive_strength, pin: D20, drvStrength: 3
```

Reply:
I see, its probably the right decision to change drive strength only for peripherals you validate. Makes sense to leave the default in other cases to the current setting.